src/uu/cp/src/cp.rs

```diff
@@ -1643,6 +1590,35 @@ fn copy_file(
         return Ok(());
     }
 
+    if options.preserve_hard_links() {
+        // if we encounter a matching device/inode pair in the source tree
+        // we can arrange to create a hard link between the corresponding names
+        // in the destination tree.
+        if let Some(new_source) = copied_files.get(
+            &FileInformation::from_path(source, options.dereference(source_in_command_line))
+                .context(format!("cannot stat {}", source.quote()))?,
+        ) {
+            // Consider the following files:
+            //
+            // * `src/f` - a regular file
+            // * `src/link` - a hard link to `src/f`
+            // * `dest/src/f` - a different regular file
+            //
+            // In this scenario, if we do `cp -a src/ dest/`, it is
+            // possible that the order of traversal causes `src/link`
+            // to get copied first (to `dest/src/link`). In that case,
+            // in order to make sure `dest/src/link` is a hard link to
+            // `dest/src/f` and `dest/src/f` has the contents of
+            // `src/f`, we delete the existing file to allow the hard
+            // linking.
+            if file_or_link_exists(dest) && file_or_link_exists(Path::new(new_source)) {
+                std::fs::remove_file(dest)?;
+            }
```

Comment:
What if we move this down to below the `handle_existing_dest` call? Wouldn't that have the same behaviour? We should probably do the proper handling for existing destination files here too right?\

So we'd move the whole preserve hard links branch to below these lines:
```
if file_or_link_exists(dest) {
    handle_existing_dest(source, dest, options, source_in_command_line)?;
}
```
or wouldn't that work?

Reply:
I don't think it will work due to this one (handle_existing_dest function) was protected by the **preserve_hardlinks**

Reply:
What do you mean with by protected? Do you mean that it wasn't called for hard links before? It does do some very specific stuff so it might indeed not work for hard links...

Reply:
Sorry for the late response,
I mean I tried to do as you said so like below
```rust
    if file_or_link_exists(dest) {
        handle_existing_dest(source, dest, options, source_in_command_line)?;
    }

    if options.preserve_hard_links() {
        // if we encounter a matching device/inode pair in the source tree
        // we can arrange to create a hard link between the corresponding names
        // in the destination tree.
        if let Some(new_source) = copied_files.get(
            &FileInformation::from_path(source, options.dereference(source_in_command_line))
                .context(format!("cannot stat {}", source.quote()))?,
        ) {
            std::fs::hard_link(new_source, dest)?;
            return Ok(());
        };
    }
```

We could not do that
due to the **handle_existing_dest** function only deleting the destination file when the overwrite option was given.

so that function wasn't to act like it's named to deal with every scenario of existing destination (if I understand that function correctly)

Reply:
But it should honor those options right? For example, if we give `-n` to (GNU) `cp` this happens:
```
❯ exa -Til
8732801 drwxr-xr-x - terts 29 aug 13:35 .
9036332 drwxr-xr-x - terts 29 aug 13:36 ├── dest
9034135 .rw-r--r-- 0 terts 29 aug 13:35 │  ├── f
9034136 .rw-r--r-- 0 terts 29 aug 13:37 │  └── g
9036331 drwxr-xr-x - terts 29 aug 13:34 └── src
9034134 .rw-r--r-- 0 terts 29 aug 13:34    ├── f
9034134 .rw-r--r-- 0 terts 29 aug 13:34    └── g

❯ cp -n --preserve=links src/f src/g dest
cp: not replacing 'dest/f'
cp: not replacing 'dest/g'
```
Note how `src/f` and `src/g` have the same inode so are hardlinks. Seems to me like the treatment of existing files is the same whether we are dealing with a second hard or not.

Reply:
Sorry, small correction the actual case that shows a difference is:
```
❯ exa -Til                                                                                                                                                                                                  13:46:18
8732801 drwxr-xr-x - terts 29 aug 13:35 .
9036332 drwxr-xr-x - terts 29 aug 13:46 ├── dest
9034136 .rw-r--r-- 0 terts 29 aug 13:45 │  └── g
9036331 drwxr-xr-x - terts 29 aug 13:34 └── src
9034134 .rw-r--r-- 0 terts 29 aug 13:34    ├── f
9034134 .rw-r--r-- 0 terts 29 aug 13:34    └── g
```
where GNU does this:
```
❯ cp -n --preserve=links src/f src/g dest
cp: not replacing 'dest/g'
```
and this PR does this:
```
❯ cp -n --preserve=links src/f src/g dest
```
(No output, `src/g` is copied)

Reply:
OK, you are right.
let me make some modifications to fulfill that goal. 🙇🏻